clients/client-java/src/main/java/com/datastrato/gravitino/client/KerberosTokenProvider.java

```diff
@@ -0,0 +1,198 @@
+  @Override
+  public void close() throws IOException {
+    try {
+      if (loginContext != null) {
+        loginContext.logout();
+      }
+    } catch (LoginException le) {
+      throw new IOException("Fail to close login context", le);
+    }
+  }
+
+  boolean isLoginTicketExpired() {
+    Set<KerberosTicket> tickets =
+        loginContext.getSubject().getPrivateCredentials(KerberosTicket.class);
+
+    if (tickets.isEmpty()) {
+      return false;
+    }
+
+    return tickets.iterator().next().getEndTime().getTime() < System.currentTimeMillis();
+  }
+
+  /**
+   * Creates a new instance of the KerberosTokenProvider.Builder
+   *
+   * @return A new instance of KerberosTokenProvider.Builder
+   */
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public static class Builder {
+    private String clientPrincipal;
+    private String serverPrincipal;
+    private File keyTabFile;
+
+    /**
+     * Sets the client principal for the HTTP token requests.
+     *
+     * @param clientPrincipal The client principal for the HTTP token requests.
+     * @return This Builder instance for method chaining.
+     */
+    public Builder withClientPrincipal(String clientPrincipal) {
+      this.clientPrincipal = clientPrincipal;
+      return this;
+    }
+
+    /**
+     * Sets the keyTabFile for the HTTP token requests.
+     *
+     * @param file The keyTabFile for the HTTP token requests.
+     * @return This Builder instance for method chaining.
+     */
+    public Builder withKeyTabFile(File file) {
+      this.keyTabFile = file;
+      return this;
+    }
+
+    /**
+     * Sets the server principal for the HTTP token requests.
+     *
+     * @param serverPrincipal The server principal for the HTTP token requests.
+     * @return This Builder instance for method chaining.
+     */
+    public Builder withServerPrincipal(String serverPrincipal) {
+      this.serverPrincipal = serverPrincipal;
+      return this;
+    }
+
+    /**
+     * Builds the instance of the KerberosTokenProvider.
+     *
+     * @return The built KerberosTokenProvider instance.
+     */
+    public KerberosTokenProvider build() {
+      KerberosTokenProvider provider = new KerberosTokenProvider();
+      Preconditions.checkArgument(
+          StringUtils.isNotBlank(clientPrincipal),
+          "KerberosTokenProvider must set clientPrincipal");
+      provider.clientPrincipal = clientPrincipal;
+
+      Preconditions.checkArgument(
+          StringUtils.isNotBlank(serverPrincipal),
+          "KerberosTokenProvider must set serverPrincipal");
+      provider.serverPrincipal = serverPrincipal;
+
+      Preconditions.checkArgument(keyTabFile != null, "KerberosTokenProvider must set keytabFile");
+      Preconditions.checkArgument(
+          keyTabFile.exists(), "KerberosTokenProvider's keytabFile doesn't exist");
+      provider.keytabFile = keyTabFile.getAbsolutePath();
```

Comment:
1. why do we need to set server principal, I don't think it is necessary.
2. we should support using tgt cache, not only keytab. Keytab is usually used for long running application.

Reply:
1. Because we use HTTP Kerberos. Client and server should co-created a context. Client should know the server principal. Otherwise, server will give client a unauthorziation exception with challenge information. client should get  server principal from challenge information of the server. This process seems more weird.
2. I will have a try.

Reply:
I implemented a tgt cache login referring to the Hadoop tgt cache implementation. 
I can't find a way to test it in the UT.
I will test it by hand when I create a kerberos environment